src/skadi/sample.cc

```diff
@@ -80,103 +68,342 @@ uint64_t file_size(const std::string& file_name) {
+#endif
+  tile_data(cache_t* c, uint16_t index, bool reusable, const int16_t* data)
+      : c(c), data(data), index(index), reusable(reusable) {
+    if (reusable) {
+      c->increment_usages(index);
+    }
+  }
+
+  tile_data(const tile_data& other) {
+    this->operator=(other);
+  }
+
+  ~tile_data() {
+    if (reusable) {
+      c->decrement_usages(index);
+    }
+  }
+
+  tile_data& operator=(const tile_data& other) {
+    c = other.c;
+    data = other.data;
+    index = other.index;
+    reusable = other.reusable;
+
+    if (reusable) {
+      c->increment_usages(index);
+    }
+    return *this;
+  }
+
+  tile_data& operator=(tile_data&& other) {
+    std::swap(c, other.c);
+    std::swap(data, other.data);
+    std::swap(index, other.index);
+    std::swap(reusable, other.reusable);
 
-const int16_t* sample::source(uint16_t index) const {
+    return *this;
+  }
+
+  inline operator bool() const {
+    return data != nullptr;
+  }
+
+  double get(double u, double v) {
+    // integer pixel
+    size_t x = std::floor(u);
+    size_t y = std::floor(v);
+
+    // coefficients
+    double u_ratio = u - x;
+    double v_ratio = v - y;
+    double u_inv = 1 - u_ratio;
+    double v_inv = 1 - v_ratio;
+    double a_coef = u_inv * v_inv;
+    double b_coef = u_ratio * v_inv;
+    double c_coef = u_inv * v_ratio;
+    double d_coef = u_ratio * v_ratio;
+
+    // values
+    double adjust = 0;
+    auto a = flip(data[y * HGT_DIM + x]);
+    auto b = flip(data[y * HGT_DIM + x + 1]);
+    if (out_of_range(a)) {
+      a_coef = 0;
+    }
+    if (out_of_range(b)) {
+      b_coef = 0;
+    }
+
+    // first part of the bilinear interpolation
+    auto value = a * a_coef + b * b_coef;
+    adjust += a_coef + b_coef;
+    // LOG_INFO('{' + std::to_string(y * HGT_DIM + x) + ',' + std::to_string(a) + '}');
+    // LOG_INFO('{' + std::to_string(y * HGT_DIM + x + 1) + ',' + std::to_string(b) + '}');
+    // only need the second part if you aren't right on the row
+    // this also protects from a corner case where you sample past the end of the image
+    if (y < HGT_DIM - 1) {
+      auto c = flip(data[(y + 1) * HGT_DIM + x]);
+      auto d = flip(data[(y + 1) * HGT_DIM + x + 1]);
+      if (out_of_range(c)) {
+        c_coef = 0;
+      }
+      if (out_of_range(d)) {
+        d_coef = 0;
+      }
+      // LOG_INFO('{' + std::to_string((y + 1) * HGT_DIM + x) + ',' + std::to_string(c) + '}');
+      // LOG_INFO('{' + std::to_string((y + 1) * HGT_DIM + x + 1) + ',' + std::to_string(d) + '}');
+      value += c * c_coef + d * d_coef;
+      adjust += c_coef + d_coef;
+    }
+    // if we are missing everything then give up
+    if (adjust == 0) {
+      return NO_DATA_VALUE;
+    }
+    // if we were missing some we need to adjust by that
+    return value / adjust;
+  }
+};
+
+tile_data cache_t::source(uint16_t index) {
   // bail if its out of bounds
   if (index >= TILE_COUNT) {
-    return nullptr;
+    return tile_data(this, index, false, nullptr);
```

Comment:
why not make a default constructor, instead of just having one for windows, that initializes the object in this way? in the case that its an invalid entry, why n ot set the cache parent to null and also set the index to some crazy value, so that every bit of the object is telling the caller this object is invalid? apologies in advance if there is an obvious reason that i didnt see

Reply:
Makes sense. There is a bool operator, to check tile state. No need to check for crazy values. :)